accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,361 @@
+definition of:
+
+- Partitions and file systems used in the node
+
+- Bootloader parameters
+
+- Repositories and packages installed
+
+- Services enabled and disabled
+
+- Users and certificates
+
+As we can see, this proposal is (1) moving closer the features of
+Kickstart based systems into the SUSE family, and (2) using SaltStack
+to orchestrate the new installation.
+
+# Detailed design
+[design]: #detailed-design
+
+## Integrating Yomi as a Salt state
+
+The core of the proposal is to add Yomi into the catalog of formulas
+available in Uyuni.
+
+As Yomi is indeed a Salt formula, this integration can be done
+directly, installing the `yomi-formula` package into the Manager.
+
+The user can later decide to add the Yomi state into the `top.sls` for
+this system, fill the pillar data using the SaltForm provided in the
+package, and launch the installer by applying the highstate.
+
+## SaltForms and sub-forms
+
+Uyuni provides a mechanism, SaltForms, that allows the creation of the
+pillar data that suit the formula. The current relation between a
+SaltForm and a Salt state is 1:1, but using some empty `init.sls` we
+have the option of splitting a big SaltForm into multiple smaller
+SaltForms (subforms), that optionally belongs to the same group.
+
+The current `yomi-formula` is using this mechanism to split the pillar
+data into six sections:
+
+- yomi: general configuration data, like use `kexec` instead of
+  `reboot` after the installation, or some `systemd-firstboot` data.
+
+- yomi-bootloader: indicate the device where the bootloader will be
+  installed, and the parameters send to the kernel.
+
+- yomi-services: list of services that will be enabled, and is
+  `salt-minion` data will be copied into the new system.
+
+- yomi-software: repository registering and packages installation.
+
+- yomi-storage: device partition, LVM or RAID configuration, and file
+  systems applied.
+
+- yomi-users: creation of initial users, passwords and certificates.
+
+As we will see in future sections, those forms contains some
+overlapping with other salt-forms (the locale configuration that
+`systemd-firstboot` requires), and with other subsystems in Uyuni
+(like the software section).
+
+## Highstate and Workflow
+
+Installing a new system will be equivalent of applying a highstate to
+a system booted from the Yomi image.
+
+The current Yomi image is a JeOS based image with a `salt-minion`,
+some systemd overwrites that will help during the initial
+configuration, and the CLI tools required when a new installation is
+performed. You can find more information here:
+https://github.com/openSUSE/yomi#the-iso-image
+
+The current Yomi states are aware of differentiating when the system
+is booting from an external image (like the Yomi image), from when the
+boot comes from the local device. This is used to decide when the
+installation of the node makes sense or not when a highstate is
+re-applied.
```

Comment:
> What happens if we mix "installation states" (e.g. partitioning) and "normal states" (e.g. adding and configuring an Apache web server) on the same system?

Uhmmm ... Do you mean mixing Yomi with other formulas? I write something about this in a different comment (https://github.com/uyuni-project/uyuni-rfc/pull/16#discussion_r332596880)

As is currently designed Yomi this is what will be happening:

* the yomi states knows about the chroot, and will provision the node and will install the packages described in the yomi pillar data inside the chroot, and will enable the services that are named inside the pillar data

* the apache-formula will install the rpm and enable the service on top on the yomi image, in the OS that is in RAM, not inside the chroot

> IMHO, the _desired_ behavior should be that you can idempotently call the high state at any point in time, and that all states in the high state are checked (and executed if required).

That is definitively desired. But I wonder how can we achieve this.

The big problem is that this 'normal state' will require to know about the chroot target, and point to it when is targeting some actions, like Yomi is doing when the software is installed:

https://github.com/openSUSE/yomi/blob/master/salt/yomi/software/pkgmanager.sls#L37

We have a module in place that help doing this transparently, that is also used by Yomi here:

https://github.com/openSUSE/yomi/blob/master/salt/yomi/users.sls#L31

Somehow, this apache formula needs to be aware that there is a chroot, and there is living.

And the relation between the apache formula and the Yomi states needs to be more deep than that. The chroot is only available during a period of time (not before the software stage and not after the post-installation state), so the apache formula needs to inject the states inside those phases.

> The only downside is that this could lead to accidental data loss, e.g. if someone changes a minor thing in a partition state (e.g. the label name of a partition) and triggers a complete re-installation by accident.
> An alternative would be to introduce the concept of a "super state" that runs "installation state" and "high state" (or even preceded by a "hardware prep" state). In that case, the "normal" high state would not include the Yomi states.

Ah yes. I think that is something like what this comment is proposing: https://github.com/uyuni-project/uyuni-rfc/pull/16#discussion_r332558510

As how the prototype is working, the pillar data and the yomi states can live during the full live cycle of the system. Yomi can detect if there is something for him or not. But you are right, in the current shape, this check enable or fully disable yomi. Internally the states are self indempotent, but if the is not rootfs to mount (because is already mounted), all the yomi states are discarded.